### ISSUE TITLE
xgboost parameter tuning(using caret)

### ISSUE THREAD

**Issue by refree** [refree/parameter-tunning#3]
#xgboost는 caret 에서 파라미터 튜닝할때 분류의 경우 목표변수 자료유형 Factor / 나머지 numeric
#진짜 xgboost에서는 목표변수 integer / 나머지 numeric

#참고 블로그(blog) https://datascience-enthusiast.com/R/ML_python_R_part2.html

setwd("C:/Users/Park Jung/Desktop/SELS/data/query")
library(caret); library(dplyr); library(randomForest); 
library(xgboost)
cb7 <- read.csv("7grade_join_1.csv" )
#cb7 <- read.csv("df.csv" )
summary(cb7$do_violence_in_school)
#- numeric --------#
for(i in 1:length(cb7)) {   cb7[,i] <- as.numeric(cb7[,i]) }
sum(is.na(cb7)); cb7 <- as.data.frame(cb7); 
str(cb7)

cb7 <- cb7 %>%
  mutate(target = ifelse(do_violence_in_school >= 0.03846, 1, 0))
str(cb7)
cb7 <- cb7[,c("school_climate", "learning_attitude",
            "pc_usage_frequency", "self_control","damage_violence_in_school",
            "self_directed_learning_attitude_behavior", "culture_activity","career_maturity","target")]

cb7$target <- as.factor(cb7$target)

library(Matrix)
#xgboost 파라미터 세팅 코드
X_train = xgb.DMatrix(as.matrix(cb7 %>% select(-target)))
y_train = cb7$target

str(X_train)

xgb_trcontrol = trainControl(method = "cv",  number = 5,  
                             allowParallel = TRUE, verboseIter = FALSE,
                             returnData = FALSE )

xgbGrid <- expand.grid(nrounds = c(100,200),  # this is n_estimators in the python code above
                       max_depth = c(10, 15, 20, 25),
                       colsample_bytree = seq(0.5, 0.9, length.out = 5),
                       ## The values below are default values in the sklearn-api. 
                       eta = 0.1, #학습률은 부스팅 단계마다 가중치를 줘서 부스팅 과정에 오버피팅이 일어나지 않도록 한
                       gamma=0, # 값 5 정보 획득량에서 로 표현한 값이며, 감마가 커질수록 트리의 깊이가 줄어들어 보수적인 모형이 된
                       min_child_weight = 1,
                       subsample = 1 )

set.seed(0)
#modelLookup(model='xgbTree')
system.time(xgb_model <- train(X_train, y_train, trControl = xgb_trcontrol, 
                               metric = "Accuracy", tuneGrid = xgbGrid,  method = "xgbTree"))
print(xgb_model); plot(xgb_model)
